<a target="_blank" href="https://colab.research.google.com/github/jgromero/dasoc-faa/blob/main/python/titanic/Sup01 - Arboles de clasificacion.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/>
</a>

<div style="text-align: center;">
    <img src="https://secretariageneral.ugr.es/sites/webugr/secretariageneral/public/inline-files/UGR-MARCA-01-color.png" alt="Logo de UGR" width="550">
</div>

# **Aplicaciones sociológicas de Inteligencia Artificial**

### Master en Ciencias de Datos aplicada a las Ciencias Sociales
#### Aplicaciones Avanzadas en Inteligencia Artificial

### Autor: Rodrigo Tapia Croes

La fecha de este informe es: **11 de junio de 2025**

<a id="Introduccion"></a>
# **Introduccion**
---

En el presente informe vamos a presentar un proyecto de aplicaciones sociologicas basadas en el uso de herramientas de Inteligencia Artificial. Las aplicaciones sociologicas se centraran exclusivamente en la investigacion y el analisis en ciencias sociales, apartando las tareas de elaboracion de proyectos sociales, el impacto etico del uso de las herramientas, la creacion de tecnicas y medidas que atenuen los efectos eticos colaterales, la transmision y transferencia de conocimientos y tecnologias, etcetera-, caracteristicas de las ciencias sociales.

De modo que, nuestro objetivo con este informe son dos:
  
  1. **Esquema programatico**: El primero, relacionado con las exigencias del master, es establecer un esquema programatico de aplicaciones de Inteligencia Artificial en el area de las ciencias sociales, y,
  2. **Interpretacion de resultados**: el otro, poner al servicio las herramientas de Inteligencia Artificial en la intepretacion de diferentes perspectivas de los datos o conjuntos de datos diponibles en cualquier investigacion.

Ambos objetivos nos plantea algunos desafios que lejos de paralizar a los cientificos sociales, ofrecen enormes potencialidades en el rol de la investigacion. Porque, nos parece que, estas herramientas mejorarian los rendimientos de la investigacion, y, a su vez, reclamaria, tareas especificas y necesarias al cientifico social.

 > En terminos generales, con la construccion del esquema programatico de investigacion, veremos que el papel del cientifico social se puede definir en la elaboracion no unicamente de los prompts, sino tambien, en la conceptualizacion y el manejo conceptual del area de especialidad. El papel del cientifico social en la conceptualizacion o el manejo conceptual del area de investigacion sera especialmente util en programar la interpretacion de resultados.

Subrayando la necesidad del cientifico social en la elaboracion de los prompsts y en la conceptualizacion, aspiramos ir cosechando algunas estrategias de la inteligencia artificial al servicio de las ciencias sociales. El presente trabajo dejara a un lado cuestiones sociales relevantes que requiere la mirada del cientifico social en la la obtencion de la informacion, la disponibilidad de la informacion, el contexto social de las fuentes de informacion, etcetera. Siendo reiterativos, nos enfocaremos exclusivamente en la inclusion programatica de los prompts y el manejo conceptual para el procesamiento e interpretacion de los datos.

De manera que, siguiendo las orientaciones del curso 2024/2025 del [*Master de Ciencias de Datos aplicada a las Ciencias Sociales*](https://www.ugr.es/estudiantes/master-doctorados/master-universitario-ciencia-datos-aplicada-ciencias-sociales), de la materia de Aplicaciones Avanzadas de Inteligencia Artificial progamaremos unas herramientas de inteligencia artificial optimizando nuestras tareas en el proceso de investigacion. El esquema programatico estada destinado a la consturccion de las siguientes tareas:

  - **Busqueda y obtencion de la informacion y la data**.

  - **Exploracion del conjunto de datos y de la estructura de los datos**.

  - **Tratamiento de los datos y la informacion**.

  - **Analisis e Interpretacion de los datos**.

En lineas generales, el esquema programatico busca construir un sistema de busqueda de informacion en repositorios en linea destinados a ser explorados y procesados por modelos de herramientas de inteligencia artificial. Agregamos una etapa de procesamiento de articulos con lo cual los agentes conceptualizaran los aspectos claves para posteriormente aplicarlos en una interpretacion de resultados. Como mencionaremos, en cada una de las etapas resaltaremos el rol del cientifico social en este proceso con aplicaciones de Inteligencia Artificial y Machine Learning.

____
** **












# **Objetivo**
---

Persiguiendo los objetivos subyacentes planteados al principio de la [Introduccion](#Introduccion), buscamos conocer las tematicas preponderantes en la opinion publica de Twitter. Para poner a prueba estos objetivos programaticos, nos planteamos un breve estudio sociologico y de marketing digital en la opinion publica de Twitter. Por ello, anexamos un objetivo subsidiario al presente informe: conocer las tendencias sociales y de marketing digital clave en la opinion de Twitter.

Las tendencia sociales y de marketing digital las conoceremos mediante una extraccion tematica de un conjunto de datos. La extraccion tematica nos permitira conocer la relacion y la influencia de determinados temas, por una parte, en la sociedad, y, por otro lado, en la movimiento de mercado.

____
** **

# **Algunas Consideraciones Teoricas**
---

Es importante que hagamos una mención previa como algunos autores sociólogos y antropólogos hacen referencia a que las emociones o las expresiones temáticas recurrentes están ligadas, también, a los contextos donde estas se manifiestan. Es decir, el sentido de un texto o de una emoción depende también del grado de relación con otras variables sociales, como puede ser el partido político, el usuario, o el tipo de canal de comunicación. Sin entrar en mayor debate, autores como Le Breton (2012) y Lively y Weed (2016) mencionan que, Clifford Geertz, Bourdieu, Blumer, Rosaldo o Hochschild consideran que el sentido social influye en la significación y manifestación de cualquier expresión individual, sean estas las expresiones emotivas o las preferencias temáticas.

Por esta razón, considerando las aportaciones teóricas de la sociología y la antropología, procederemos a ampliar nuestras técnicas de Inteligencia Artificial para explotar las posibilidades que nos ofrecen estas técnicas para conocer las emociones y los temas contextualmente localizados.

Principalmente, por las potencialidades de los modelos de Inteligencia Artificial para flexibilizar sus análisis e interpretaciones sobre la actuación en grandes conjuntos de datos.


** **
___


<a id="Metodologia"></a>
# **Metodologia**
---
  
Para satisfacer este objetivo de investigacion, y los objetivos programaticos consustanciales, nos plantearemos una investigacion mediante la siguiente metodologia: la busqueda de un conjunto de datos en un repositorio de datos, para luego de su exploracion, hacer el procesamiento de extraccion tematica destinada a la interpretacion de los resultados desde la perspectiva sociologica y de marketing digital.

 > La investigacion sera una breve aproximacion sobre el sentido que tienen algunas de las tematicas de la opinion publica de Twitter. El sentido que aspiramos conocer seran desde la perspectiva del marketing digital y la sociologia. Pero, obviamente, nuestra labor se concentrara en ofrecer un banco conceptual a un agente de Inteligencia Artificial para que interprete los resultados procesados por modelos de Inteligencia Artificial orientados por nuestros prompts.

Siendo mas precisos, nuestra metodologia en cada una de las etapas fue la siguiente:

  1. Obtencion y busqueda de informacion:
     - Programamos el [Large language model] **`Gemini 1.5 Flash`** para la busqueda de conjuntos de datos de Twitter basados en opiniones publicas y generales, especificando el repositorio [Kaggle](https://www.kaggle.com).
     - Luego de esta busqueda, incluimos una busqueda de meta datos del conjunto de datos seleccionados entre las disintas alternativas.     
     - No obstante, a pesar de nuestra eleccion inicial -[Brexit twitter data](https://www.kaggle.com/datasets/ironbatshashank/brexit-twitter-data)-, combinar esta herramienta con tecnicas de busqueda de machine learning de [DuckDuckGo-Search](https://pypi.org/project/duckduckgo-search/) permitio conocer los enlaces de las alternativas de busqueda ofrecidas por el modelo. Buscando ampliar nuestro objetivo de investigacion, seleccionamos -[Tweets Datasets](https://www.kaggle.com/datasets/mmmarchetti/tweets-dataset)- que ofrecia opiniones mas variadas y diversas.
     - Este proceso los concluimos con la descarga y el almacenamiento del conjunto de datos en la nube de **Google Colab**, y su posterior conversion en un data frame.

2. Exploracion del conjunto de datos y de la estructura de los datos:
   - El almacenamiento del conjunto de datos en un data frame, permitio una exploracion exhaustiva de los datos contenidos. La exploracion de datos se ejecuto con el modelo **`"llama3-70b-8192"`** mediante [Groq](https://console.groq.com/home). Para generar una serie de respuestas sobre la exploracion de datos conectamos el servicio interactivo de **Llama** a travez de [LangChain](https://www.langchain.com/). Con este agente, conocimos las dimensiones, las variables, el tipo de objeto de las variables, los valores unicos de las variables, la presencia de valores duplicados, los valores nulos, y algunas medidas para los valores numericos como la media, la mediana y la desviacion estandar.
  - La exploracion de datos nos permitio conocer las variables relevantes para el analisis de influencia tematica desde las perspectivas propuestas (sociologica y marketing digital). Las variables `"number_of_likes"`, `"number_of_shares"` necesarias para conocer la influencia de los temas subyacentes en la vaiable `"content"`.
    - Luego, haciendo uso de un modelo de [Antropic](https://www.anthropic.com/api), realizamos la extracción de temas de `"content"` hallando los principales tópicos en la opinión pública. El uso de [Antropic](https://www.anthropic.com/api) añade una estrategia para captar la relación contextual de esos tópicos, que llamaremos ‘el contexto temático’. Hicimos una extraccion tematica por registro mediante el modelo **`"claude-3-haiku-20240307"`**.
  - El resultado fue un data frame con otra variable `"analisys_results"`. La nueva variable contenia una lista con los tres temas, una lista con los hashtags y otra lista con urls mas relvantes. En otras palabras, la nueva variable tenia un registro de tres lista que requirio una expansion de las listas en nuevas columnas, diferenciando nuestra categoria `"temas"` con los topicos importantes solicitados.
  - La informacion la guardamos en un documento *csv* para su posterior analisis e interpretacion.
  3. **Tratamiento de los datos y la informacion**
  - Otra vez haciendo uso de [Groq](https://console.groq.com/home), pero esta vez del modelo **`"llama3-8b-8192"`** realizamos un primer analisis de la informacion. Nuestro primer analisis consistio en concoer la influencia de los temas principales, mediante su relacion con los likes y shares. Con esta caracterizacion de los temas principales guardamos los resultados de los numeros de likes y shares de cada tema en un documento **`"llm_responses_engagement_analysis.txt"`**.
  - Este analisis nos permitio agregar una influencia de los temas en el ecosistema de opinion publica disponible en el conjunto de datos, con lo cual podriamos refinar la interpretacion de nuestros resultados.

  4. **Analisis e Interpretacion de los datos**.
   -  En esta etapa crucial, el análisis y la interpretación de los resultados se llevaron a cabo mediante dos fases diferenciadas, pero interconectadas:
    
  *a) La creacion del marco de referencia conceptual*:
      
      - **Búsqueda y Criterios de Artículos Científicos**: Iniciamos un proceso de búsqueda de artículos científicos y especializados en sociología y marketing digital. La selección de estos documentos se basó en criterios de simplicidad, precisión y operacionalización de los conceptos, lo cual es fundamental para optimizar el trabajo posterior con modelos de lenguaje generativo y sistemas de Generación Aumentada por Recuperación (RAG).
       - **Procesamiento Conceptual con Mistral AI y LlamaIndex**: Tras la selección, realizamos un procesamiento avanzado de los documentos para extraer y definir los conceptos clave utilizando los modelos de [Mistral](https://docs.mistral.ai/api/.). Esta tarea se ejecutó de la siguiente forma:
            - *Vectorización de Embeddings para Documentos*: Dado que trabajamos con documentos PDF, empleamos el modelo de embedding **`"mistral-embed"`**  para generar embeddings vectoriales de cada documento seleccionado. Este proceso transformó el contenido textual de los PDFs en representaciones numéricas de alta dimensión. Estos embeddings son cruciales porque capturan el significado semántico de las palabras y frases, permitiendo que los modelos de lenguaje comprendan y procesen el contexto de los artículos para la posterior extracción de conceptos clave solicitados en nuestros prompts.

            - *Caracterización Conceptual con LlamaIndex*: La caracterización y definición de los conceptos a partir de los artículos científicos se realizó utilizando el modelo **`"mistral-large-latest"`**. Aquí, [LlamaIndex](https://www.llamaindex.ai/) jugó un papel central como framework de orquestación y gestión de datos para LLMs. LlamaIndex nos permitió indexar los documentos PDF (a partir de sus embeddings) en una base de datos vectorial. Esto facilitó la recuperación eficiente de fragmentos relevantes de los artículos en respuesta a nuestras consultas, lo que, en interacción con **`"mistral-large-latest"`**, posibilitó la generación de definiciones precisas y la extracción de conocimientos pertinentes.

            - *Extracción y Registro de Definiciones*: Finalmente, extrajimos las definiciones clave esenciales para las perspectivas sociológica y de marketing digital, registrándolas en un documento de texto (**`"informacion_extraida_de_pdfs.txt"`**).

  *b) La interpretacion de la influencia de los temas*:

     - **Cadena RAG Orientada por Conceptos con LangChain**: En base a los conceptos y resúmenes clave obtenidos en la fase anterior, configuramos una cadena RAG (Generación Aumentada por Recuperación) utilizando [LangChain](https://www.langchain.com/). A esta cadena RAG se le suministraron las orientaciones interpretativas necesarias, basadas en nuestro marco conceptual, para analizar y contextualizar los resultados de nuestro estudio.
     - **Embeddings para Consultas y Contexto**: Para la operación efectiva de la cadena RAG, fue indispensable generar los embeddings de nuestras consultas (las preguntas o solicitudes de interpretación que hacíamos al modelo) y del contexto relevante (fragmentos de nuestros resultados de análisis y los conceptos del marco conceptual). Para esta tarea, empleamos el modelo de embedding **`"models/embedding-001"`**. Estos embeddings permitieron a la cadena RAG identificar y recuperar la información más pertinente para cada solicitud.
     - **Modelo de Interpretación**: La fase de interpretación se llevó a cabo con el modelo **`"gemini-1.5-flash-latest"`**. Dentro de la cadena RAG orquestada por [LangChain](https://www.langchain.com/), este modelo recibía los conceptos clave previamente definidos (en el marco_conceptual, **`"informacion_extraida_de_pdfs.txt"`**). De esta manera, el modelo interpretó los resultados registrados en nuestro documento de análisis (e.g.,**`"llm_responses_engagement_analysis.txt"`**), a la luz de los conceptos definidos en el paso previo (**`"informacion_extraida_de_pdfs.txt"`**). Este proceso garantiza una interpretación guiada y basada en el conocimiento específico del dominio.

Siguiendo este esquema programatico, establecimos una interpretación contextualizada y guiada por las definiciones relevantes de nuestro banco conceptual. En este esquema programatico de investigacion social y de marketing digital, puntualizamos la importancia del cientifico social en la seleccion de las variables, la creacion de los prompts, la configuración de la cadena LLM (Conexión y Ejecución) el manejo y eleccion del marco referecial -con el que los modelos de Inteligencia Artificial elaboraran las definiciones, y con ello la posterior interpretacion.  

## **Herramientas utilizadas**
---

Todo el trabajo se realizó en Google Colab, lo cual facilitó el uso de Python y librerías de IA en la nube. Utilizamos herramientas como Gemini, Langchain, LlamaIndex, Groq, Mistral, Antropic y DuckDuckGo Search, entre otras.

Google Colab facilitó el acceso a infraestructura en la nube, permitiendo la integración de diversas librerías sin necesidad de configuraciones complejas en una máquina local. Esta elección permitió el acceso a GPU y TPU sin mayor coste, la integración de librerías como Langchain, pandas, Langchain-groq, Langchain-mistral, Langchain-gemini de manera rápida; y trabajar en diversas sesiones al mismo tiempo.

En cuanto a la justificación de las herramientas de inteligencia artificial decidimos hacer una división de su uso para tareas especializadas por cada herramienta. Una vez definidos los procedimientos metodológicos para el análisis de texto, los criterios de selección de las herramientas para las tareas fueron los siguientes:

1. **Groq para el análisis**: Groq es conocido por su enfoque en inteligencia artificial de alta velocidad y bajo consumo. Esta herramienta ofrece eficiencia en el análisis de datos o texto con tiempos de respuesta ultrarrápidos, por eso, Groq puede fue una excelente opción. Su arquitectura optimizada lo hace ideal para tareas en tiempo real tanto en el analisis exploratorio como en el analisis descriptivo de los temas, numeros de likes y numeros de shares.

2. **Mistral para la generacion de resumenes y conceptos claves**:Su versatilidad y fácil implementación lo hacen una herramienta accesible y poderosa para clasificar emociones en texto. A pesar de las exigencias de las salidas en formato json o txt, mistral ofrece una capacidad de vectorizacion de textos y generaxion de texto. Esas capacidades posiciona bien los modelos de generacion de texto de Mistral para destilar información compleja.

3. **Antropic para extracción de temas**: Antropic Cloud AI ha desarrollado modelos compactos y eficientes que sobresalen en tareas como la extracción de tópicos clave. Los modelos de Anthropic son utiles para identificar patrones y temas principales en grandes volúmenes de datos, siendo una aplicación directa de sus fortalezas.

4. **Gemini para la generacion de texto y la busqueda de informacion**: Gemini de Google es poderoso en procesamiento de lenguaje y búsqueda contextual. Su capacidad para entender y generar contenido basado en búsquedas hace que sea muy útil cuando necesitas información de múltiples fuentes y una visión amplia de un tema. Por esta razon, Gemmini resulta ser la herramienta ideal para la interpretacion de los analisis obtenidos.

5. **LangChain para la interaccion con los modelos de lenguaje**: LangChain se enfoca en la orquestación y la lógica de aplicación. Su fortaleza es construir cadenas de razonamiento complejas, integrar LLMs con herramientas externas y gestionar el flujo de la interacción.

  Por esa razon, LangChain fue seleccionado como el framework orquestador central de este proyecto por su capacidad para facilitar la construcción de aplicaciones complejas impulsadas por modelos de lenguaje (LLMs). En el contexto de un proyecto de investigación sociológica que integra múltiples LLMs y diversas etapas de procesamiento de datos, LangChain resultó indispensable por las siguientes razones:

  Proporciono una interfaz unificada para la interaccion con los diferentes modelos (Groq, Mistral, Anthropic, Gemini) para tareas especializadas. A su vez, sirvio para la integracion de herramientas para la manipulacion de los datos y la informacion ordenadas.

  Permitio la construcción de cadenas y agentes de IA. Esta funcionalidad es crucial para nuestro esquema programático, ya que nos permite encadenar las tareas de "búsqueda y obtención de información", "exploración", "tratamiento" y "análisis/interpretación". Por ejemplo, podemos construir una cadena donde un LLM procesa un texto, otro lo resume y un agente final interpreta los resultados basándose en conocimientos externos.

  Facilito la creación, gestión y optimización de prompts complejos, asegurando que las instrucciones a los LLMs sean claras y que sus respuestas puedan ser parseadas y utilizadas eficientemente en etapas posteriores del análisis.


  6. **LlamaIndex para la convergencia y manejo de documentos pdf**: fue la herramienta clave para construir y gestionar nuestra base de conocimiento personalizada, permitiendo que nuestros agentes de IA accedieran a la información relevante para un análisis y una interpretación profund basada en datos. La principal fortaleza de LlamaIndex radica en la ingesta, indexación y recuperación eficiente de datos no estructurados (como documentos PDF) para hacerlos "comprensibles" y "buscables" por los modelos de lenguaje. Esta fortaleza nos facilito la creación de índices optimizados (como índices vectoriales) a partir de nuestros datos.

  Al integrar la capacidad de recuperación con los LLMs, la indexacion de la similitud semantica rapida y precisa hizo susceptiple las cadenas RAG de consulta para nuestra base de conocimientos específica (los datos de la investigación). La creacion de esas cadenas RAG basada en los conocimientos especificos fue relevante en la recuperacion de los fragmentos de información  antes de la generacion de una respuesta. De modo que, LlamaIndex hizo posible interpretaciones contextualmente relevantes y verificables de nuestros propios datos, que en este caso fueron documentos externos.

7. **Fuente de Datos Abiertos y Diversos de Kaggle**: Kaggle es uno de los repositorios más grandes y accesibles de conjuntos de datos públicos. Para la etapa "a) Búsqueda y obtención de la información y la data", Kaggle ofrece una vasta colección de datasets relevantes para las ciencias sociales (demografía, encuestas, redes sociales, comportamiento, etc.), facilitando el acceso a datos para la experimentación y el desarrollo de nuestro esquema programático.

  Los conjuntos de datos en Kaggle a menudo vienen acompañados de descripciones detalladas, discusiones en foros y análisis de otros usuarios. Esta información contextual es valiosa para comprender mejor la naturaleza y las posibles interpretaciones de los datos antes de aplicar nuestros modelos de IA.

8. **Plataformas y IA en la nube**

  Además de las librerías y frameworks, las plataformas de desarrollo de los propios proveedores de LLMs fueron vitales para la fase de experimentación y optimización. Al ser servicios de IA basados en la nube, plataformas como Google AI Studio (para Gemini), el Workbench de Anthropic y la plataforma de Mistral AI ofrecieron entornos interactivos donde pudimos testear y refinar prompts complejos en tiempo real.

  Estas interfaces no solo facilitaron la consulta y generación de respuestas iniciales, sino que también proporcionaron la base para la construcción de códigos especializados de interacción con sus APIs, agilizando significativamente la integración de sus modelos en nuestro pipeline de análisis e interpretación.
_____
** **

# **Esquema programatico**

## **1. Busqueda y obtencion de datos**

---




  La primera de las tareas fue realizar una busqueda exhaustiva de los conjuntos de datos disponible en [Kaggle](https://www.kaggle.com). Pero, como hemos hecho enfasis, desarrollamos una aplicacion en Inteligencia Artificial para hacer la busqueda de un conjunto de datos segun un tema y repositorio especifico. Con la inclusion de dos inputs del tema y el repositorio del conjuunto de datos logramos reducir las opciones de nuestra busqueda y, con ello, la accesibilidad del data frame. Para ello, la busqueda se centro en confirmar el autor, el repositorio, el año, el titulo del data set, ademas de un identificador que incluimo para el acceder y almacenar los criterios de busqueda.

  La segunda de las taareas fue la exploracion de la meta data de la informacion del conjunto de datos seleccionados. La inclusion del identificador, permitio crear un modelo de exploracion de la meta data disponible en la pagina web. Es importante mencionar, que esta consulta era una exploracion que se debia corroborar con analisis exploratorios posteriores en el procesamiento de los datos.

  La tercera de la tarea fue incluir un programa de aprendizaje autmatico con DuckDuckGo-search, para la busqueda de los enlaces que nos permitirian descargar los conjuntos de datos seleccionados.



### **Primera Tarea: configuracion de un modelo de busqueda de data sets**

  Como mencionamos en la [metodologia](Metodologia) este paso fue el resultado de la configuracion de un modelo de busqueda de datos en repositorios de datos con **`Gemini 1.5 Flash`**. El modelo se construyo a partir de dos estructuras basicas: a) la creacion de la plantilla de instrucciones (la elaboracion del prompt) y, b) la configuracion de la cadena RAG (conexion y ejeccion).
  
  a) En cuando a la creacion de la plantilla de instruciones definimos la operatividad del modelo como un buscador de conjuntos de datos de un tema y un repositorio especifico. Para la configuracion de un modelo de busqueda, la plantilla de instrucciones del modelo fue la siguiente:

   > " Encuentra datasets de datos abierotos y públicos relacionados con esta búsqueda: `{consulta}`. Los datasetes deben ser descargables en este repositorio `{repositorio}`. Ademas, le pedimos una devolucion de los resultados en una estructura JSON con los siguientes items:

        - "ID": 1,
        - "repositorio": "Kaggle",
        - "titulo": "Nombre del dataset",
        - "autor": "Autor",
        - "año": "2023"

La idea de este modelo de busqueda de conjuntos de datos tenia la finalidad de proporcionar los items basicos necesarios para la obtencion de esos data sets a traves de cualquier buscador de internet. Asignar un identificador para poder acceder a los conjuntos de datos encontrados, y, por supuesto, con el nombre del repositorio, titulo, autor y año.

Una vez creada la plantilla de instruccion, configuramos la cadena de RAG para optimizar la conexion y ejecucion del modelo con respecto a la busqueda de los data sets. En este caso, la configuracion del RAG se hizo mediante la incorporacion de los inputs de `{consulta}` y `{repositorio}` con las siguientes preguntas

> consulta = input("🔍 ¿Qué tipo de dataset estás buscando?: ")

> repositorio = input("🔍 ¿En que repositorio estás buscando?: ")

Finalmente, los resultados serian almacenados en un vector.

De manera que, los pasos del modelo de busqueda de informacion fue:
- La configuracion de la API key de Google AI studio (Gemini).
- La creacion del modelo Gemini.
- La creacion de la funcion basada en una interaccion conversacional con el modelo (la plantilla de instrucciones y la configuracion de la cadena RAG).
- Almacenamiento de los resultados.

#### Creacion del modelo de Gemini

In [ ]:
# ✅ Paso 1: Instalacion de la libreria necesaria
!pip install -q google-generativeai


In [ ]:
# ✅ Paso 2: Configurar tu API Key de Google AI Studio (Gemini)
import google.generativeai as genai

In [ ]:
# Reemplaza con tu propia clave
genai.configure(api_key="GOOGLE_API_KEY")
# ✅ Paso 3: Crear modelo de Gemini
modelo = genai.GenerativeModel("gemini-1.5-flash")

#### Creacion de la funcion basada en una interaccion conversacional con el modelo.

In [ ]:
import re
import json # Import the json module
def buscar_datasets_con_ia(consulta):
    prompt = f"""
    Encuentra datasets de datos abierotos y públicos relacionados con esta búsqueda: '{consulta}'.
    Los datasetes deben ser descargables en este repositorio '{repositorio}'.
    Devuélveme solo una lista JSON con máximo 10 resultados con esta estructura:
    [
      {{
        "ID": 1,
        "repositorio": "Kaggle",
        "titulo": "Nombre del dataset",
        "autor": "Autor",
        "año": "2023"
      }},
      ...
    ]
    """ # Se construye la pregunta especifica 'el prompt' con el que se va a interactuar con el modelo

    respuesta = modelo.generate_content(prompt) # Se define la interaccion con el modelo
    texto = respuesta.text # Extraccion del texto con la respuesta Gemini
    json_match = re.search(r"\[.*\]", texto, re.DOTALL) # El resto logica de extraccion y procesamiento JSON
    if json_match:
        try:
            return json.loads(json_match.group(0))
        except json.JSONDecodeError:
            print("⚠️ Error al decodificar JSON:")
            print(texto)
            return []
    else:
        print("⚠️ No se encontró JSON:")
        print(texto)
        return []

# --- Ejecutar búsqueda ---
consulta = input("🔍 ¿Qué tipo de dataset estás buscando?: ") # Especificar la primera condicion de la consulta
repositorio = input("🔍 ¿En que repositorio estás buscando?: ") # Especificar la segunda condicion de la consulta dos
datasets = buscar_datasets_con_ia(consulta) # Almacenamiento de los resultados con la interaccion de Gemini

if not datasets:
    print("❌ No se encontraron resultados.")
else:
    print("\n✅ Resultados encontrados:")
    for ds in datasets:
        print(f"[{ds['ID']}] {ds['titulo']} ({ds['repositorio']}, {ds['año']}) - Autor: {ds['autor']}")

    # Guardar en archivo para usar luego
    with open("resultados_datasets.json", "w", encoding="utf-8") as f:
        json.dump(datasets, f, indent=2, ensure_ascii=False)
    print("\n📁 Resultados guardados en 'resultados_datasets.json'")

🔍 ¿Qué tipo de dataset estás buscando?: Tweets datasets politics
🔍 ¿En que repositorio estás buscando?: Kaggle

✅ Resultados encontrados:
[1] Tweets about Politics from 2008-2019 (Kaggle, 2019) - Autor: Unknown (Often aggregated from various sources)
[2] Twitter US Airline Sentiment (Kaggle, 2015) - Autor: Kayla Skinner
[3] Sentiment140 (Kaggle, 2010) - Autor: Stanford
[4] Brexit Tweets Dataset (Kaggle, 2016) - Autor: Various Contributors
[5] Crowdflower Data:  Sentiment Analysis for Twitter (Kaggle, 2013) - Autor: Crowdflower
[6] Twitter data for analyzing public opinion (Kaggle, 2020) - Autor: Various Contributors
[7] Disaster Tweets (Kaggle, 2018) - Autor: Abdul Majed
[8] COVID-19 Tweets Dataset (Kaggle, 2020) - Autor: Various Contributors
[9] Elon Musk Tweets (Kaggle, 2023) - Autor: Various Contributors
[10] Election Tweets Dataset (Kaggle, 2020) - Autor: Various Contributors

📁 Resultados guardados en 'resultados_datasets.json'


---
#### Resumen
Este modelo proporciona una base sólida para la obtención programática de conjuntos de datos, un paso crucial en cualquier investigación basada en datos dentro de las ciencias sociales. Como se puede ver los resultados fueron encontrados con los items esperados, y asignandole un identificador.

### **Segunda Tarea: exploracion de la metada en la ubicacion del conjunto de datos en el repositorio**

 Posteriormente a la busqueda de informacion, siguiendo nuestras indicaciones en la [metodologia](Metodologia) realizamos otra configuracion del **`Gemini 1.5 Flash`**. Esta vez el modelo definia la creacion de plantillas de instrucciones para la exploracion de la meta data disponible en Kaggle, asi como, una configuracion RAG a conveniencia de esas instrucciones.
  
  a) En cuando a la creacion de la plantilla de instruciones como ya poseiamos la informacion especifica para el hallazgo del data set, incluimos en el prompt inicial esas variables identificadas en la busqueda previa. Asimismo, la operatividad del modelo como un buscador de conjuntos de datos de un tema y un repositorio especifico implico la agregacion de otras variables, muchas veces publicas en [Kaggle](https://www.kaggle.com).
  
  Nuestro principal interes era reconocer no solamente los criterios de identificacion del data set, sino tambien el numero de variables, registros, y los tipos de variables. Ademas, agregamos que el modelo hiciera un comentario y observacion del conjunto de datos buscado.

   > "  Eres un asistente experto en datasets públicos. Basado en el dataset titulado `{titulo}`, del repositorio `{repositorio}`, creado por `{autor}` en el año `{año}`,
    realiza un análisis estimado de la metadata interna del dataset y una breve descripción general (resumen) del contenido. A su vez, especificamos la estructura dell contenido hallado como se muestra a continuacion:

    Devuelve un JSON con esta estructura:
    {{
      "titulo": "{titulo}",
      "repositorio": "{repositorio}",
      "autor": "{autor}",
      "año": "{año}",
      "resumen": "breve resumen descriptivo del dataset",
      "numero_registros": número entero estimado,
      "numero_variables": número entero estimado,
      "nombres_variables": [lista de strings con nombres de variables],
      "tipos_variables": [lista de strings con tipos de variables (ej: numérico, texto, fecha)],
      "comentarios": "observaciones adicionales relevantes del dataset"
    }}

    SOLO responde con el JSON.

En este caso, es importante mencionar que, el modelo era un buscador de meta data exploratorio, que, hallaba datos publicados, aproximandose a la informacion disponible. Muchas veces, cuando al informacion expuesta en la pagina de estos data sets excede a un solo conjunto de datos, el buscador simplificaba la informacion, como es el caso del numero de registros o las variables, ofreciendo una aproximacion.

De alli, la necesidad de aplicar los analisis exploratorios propiamente a los conjuntos de datos, luego de la obtencion de los enlaces o el almacenamiento de los data sets.

Como en el anterior modelo, una vez creada la plantilla de instruccion, configuramos la cadena de RAG para optimizar la conexion y ejecucion del modelo con respecto a la busqueda de la meta data. En este caso, agregamos un input sobre el identificador perteneciente a la lista de data  frames, como se expone a continuacion:

> id_elegido = int(input("\n🔢 Ingresa el ID del dataset para analizar su metadata: "))

> seleccionado = next((d for d in datasets if d["ID"] == id_elegido), None)
Finalmente, los resultados serian almacenados en un vector.

Es importante señalar que, la exploracion de la meta data ofrecia un caracter exploratorio inicial para la eleccion del conjunto de datos mas optimos para los objetivos de una investigacion.

Coherentemente a la creacion de un modelo de IA, los pasos fueron:
- La configuracion de la API key de Google AI studio (Gemini).
- La creacion del modelo Gemini.
- La creacion de la funcion basada en una interaccion conversacional con el modelo (la plantilla de instrucciones y la configuracion de la cadena RAG).
- Almacenamiento de los resultados.

#### Configuracion del modelo Gemini

In [ ]:
# Configura tu API Key de Google AI Studio (Gemini)
genai.configure(api_key="GOOGLE_API_KEY")

# Crea el modelo de Gemini (puedes usar el mismo modelo)
modelo_metadata = genai.GenerativeModel("gemini-1.5-flash")

#### Creacion de la funcion para la consulta de la meta data

In [ ]:
def analizar_metadata_con_ia(autor, repositorio, titulo, año):
    prompt = f"""
    Eres un asistente experto en datasets públicos.
    Basado en el dataset titulado '{titulo}', del repositorio '{repositorio}', creado por '{autor}' en el año {año},
    realiza un análisis estimado de la metadata interna del dataset y una breve descripción general (resumen) del contenido.

    Devuelve un JSON con esta estructura:
    {{
      "titulo": "{titulo}",
      "repositorio": "{repositorio}",
      "autor": "{autor}",
      "año": "{año}",
      "resumen": "breve resumen descriptivo del dataset",
      "numero_registros": número entero estimado,
      "numero_variables": número entero estimado,
      "nombres_variables": [lista de strings con nombres de variables],
      "tipos_variables": [lista de strings con tipos de variables (ej: numérico, texto, fecha)],
      "comentarios": "observaciones adicionales relevantes del dataset"
    }}

    SOLO responde con el JSON.
    """ # Otra vez pedirle a Gemini la estimacion de la metadata de un data set, agregando algunos itmes claves, entre ellos, la asignacion de un identificador.
    respuesta = modelo_metadata.generate_content(prompt) # Se define la interaccion con el modelo
    texto = respuesta.text # Extraccion del texto con la respuesta Gemini
    json_match = re.search(r"\{.*\}", texto, re.DOTALL)# El resto logica de extraccion y procesamiento JSON
    if json_match:
        try:
            return json.loads(json_match.group(0))
        except json.JSONDecodeError:
            print("⚠️ Error al decodificar JSON:")
            print(texto)
            return None
    else:
        print("⚠️ No se encontró JSON en la respuesta:")
        print(texto)
        return None

# --- Cargar resultados previos ---
try:
    with open("resultados_datasets.json", "r", encoding="utf-8") as f:
        datasets = json.load(f)
except FileNotFoundError:
    print("❌ No se encontró 'resultados_datasets.json'. Ejecuta primero el flujo de búsqueda.")
    exit()

# --- Selección por ID ---
id_elegido = int(input("\n🔢 Ingresa el ID del dataset para analizar su metadata: "))
seleccionado = next((d for d in datasets if d["ID"] == id_elegido), None)

if seleccionado:
    print("\n🔍 Consultando metadata estimada para:")
    print(f"📌 Título: {seleccionado['titulo']}")
    print(f"📁 Repositorio: {seleccionado['repositorio']}")
    print(f"👤 Autor: {seleccionado['autor']}")
    print(f"📅 Año: {seleccionado['año']}")

    metadata = analizar_metadata_con_ia(
        autor=seleccionado["autor"],
        repositorio=seleccionado["repositorio"],
        titulo=seleccionado["titulo"],
        año=seleccionado["año"]
    )

    if metadata:
        print("\n📊 Metadata estimada:")
        print(json.dumps(metadata, indent=2, ensure_ascii=False))
    else:
        print("❌ No se pudo analizar la metadata.")
else:
    print("❌ ID no encontrado.")



🔢 Ingresa el ID del dataset para analizar su metadata: 4

🔍 Consultando metadata estimada para:
📌 Título: Brexit Tweets Dataset
📁 Repositorio: Kaggle
👤 Autor: Various Contributors
📅 Año: 2016

📊 Metadata estimada:
{
  "titulo": "Brexit Tweets Dataset",
  "repositorio": "Kaggle",
  "autor": "Various Contributors",
  "año": "2016",
  "resumen": "Este dataset contiene una colección de tweets relacionados con el referéndum del Brexit en el Reino Unido de 2016.  Probablemente incluye información sobre las opiniones expresadas, las tendencias y la actividad en las redes sociales durante ese periodo crucial.  La información se centra en la discusión pública sobre la salida del Reino Unido de la Unión Europea.",
  "numero_registros": 100000,
  "numero_variables": 5,
  "nombres_variables": [
    "tweet_id",
    "created_at",
    "text",
    "user_id",
    "sentiment"
  ],
  "tipos_variables": [
    "numérico",
    "fecha",
    "texto",
    "numérico",
    "texto"
  ],
  "comentarios": "El núme

___

#### Resumen

Con la informacion disponible vemos, por ejemplo, un conjunto de datos de Kaggle sobre la coleccion de comentarios acerca del brexit. El modelo nos ofrece una aproximacion del numero de variables, registros, los nombres (de los cuales podemos deducir las variables), y el tipo de variables. Con esta informacion podemos saber si el conjunto de datos es susceptible de un analisis tematico, o de otro tipo, de acuerdo a las finalidades de la investigacion.

De modo que, luego de la exploracion de la meta data de todos los conjuntos de datos, pudimos seleccionar los datos mas conveniente en el repositorio escogido. La eleccion del repositorio escogido se debe a la confianza y a la credibilidad de este repositorio, que ademas cuenta con una gran comunidad de usuarios.


Luego de observar esta descripcion exploratoria del conjunto de datos escogido podemos reconocer que sirve, en nuestro caso, para un analisis de temas desde las perspectivas de la sociologia y del marketing digital.



### **Tercera Tarea: Busqueda de los enlaces de los conjuntos de datos**

  Con la informacion de los conjuntos de datos en la disposicion de nuestros objetivos de estudios, podemos realizar unas tareas automaticas para la busqueda de los enlaces de los data sets identificados, y su posterior descarga. En este caso, hicimos una tarea de busqueda de la libreria DuckDuck-Search, optima para e descubrimiento de informacion web. Ademas, la libreria, al estar estructurada con el buscador, permite que obtengamos los enlaces de descarga de los datasets. La obtencion de estos enlaces es posible por los criterios de busquedas y los items identificadores de los conjuntos de datos.

  El bloque de código toma el dataset que Gemini identificó, construye una búsqueda precisa para DuckDuckGo usando la información de ese dataset, presenta los enlaces encontrados al usuario y le permite seleccionar el enlace real desde donde desea descargar los datos. Es un paso crucial para pasar de la identificación de un dataset a su obtención real.

  Los pasos del codigo programa las siguientes tareas:

  - Seleccion del data set: de acuerdo al identificador de los conjuntos de datos se introduce el numero esperado.
  - Construccion de la consulta: a traves de los items resguardados en la lista de los conjuntos de datos se constrye la instruccion de busqueda.
  - Ejecucion de la busqueda: la busqueda se realiza a partir de los criterios identificadores del nombre del conjunto de datos y el repositorio, año y autor.
  - Presentacion de los enlaces: a pesar de que la busqueda esta restringida, es necesario mencionar que al ampliar el numero de enlaces de busqueda, el comando ofrece las alternativas cercanas a la informacion de la consulta obtenida. Es decir, el resultado no solo es el conjunto de datos esperados, sino aquellos conjuntos de datos que se aproximen a los criterios de busqueda. Este paso es relevante, porque nos permite explorar manualmente el conjunto de datos, y, hacer la descarga de acuerdo a nuestros objetivos de estudio, sin restringirse a la eleccion inicial del data set.

  De hecho, nuestra eleccion inicial de un data set acerca de comentarios en Twitter sobre el brexit, extendio la presentacion con otros conjuntos de datos, que nos resultaron mas interesantes, y entre los cuales nos decidimos por uno sobre discusiones generales en Twitter.

In [ ]:
# ✅ Paso 1: Instalar dependencias necesarias
!pip install -q duckduckgo-search pandas google-generativeai

In [ ]:
from duckduckgo_search import DDGS

if datasets:
    try:
        # Use the previously defined variable id_elegido instead of the undefined id_consulta
        dataset_seleccionado = next((item for item in datasets if item.get("ID") == id_elegido), None)

        if dataset_seleccionado:

            consulta_busqueda = f"site:{dataset_seleccionado.get('repositorio').lower()}.com {dataset_seleccionado.get('titulo')} dataset"

            print(f"\nBuscando enlaces reales con DuckDuckGo para: {consulta_busqueda}")

            with DDGS() as ddgs:
                resultados_ddg = list(ddgs.text(consulta_busqueda, max_results=10))

            if resultados_ddg:
                print("\nEnlaces encontrados:")
                for i, resultado in enumerate(resultados_ddg):
                    print(f"[{i}] {resultado['href']}")

                # Solo pedir que selecciones el índice del enlace que quieres usar
                id_enlace = int(input("\nSelecciona el enlace (por número) donde quieres descargar los datos: "))
                if 0 <= id_enlace < len(resultados_ddg):
                    print(f"\nSeleccionaste el enlace: {resultados_ddg[id_enlace]['href']}")
                else:
                    print("Número de enlace inválido.")
            else:
                print("No se encontraron enlaces con DuckDuckGo.")

        else:
            print("No se encontró ningún dataset con ese ID.")
    except ValueError:

        print("Por favor, ingresa un número entero válido.")
else:
    print("No hay datasets cargados.")


Buscando enlaces reales con DuckDuckGo para: site:kaggle.com Brexit Tweets Dataset dataset

Enlaces encontrados:
[0] https://www.kaggle.com/datasets/visalakshiiyer/twitter-data-brexit
[1] https://www.kaggle.com/datasets/prakashpraba/twitter-sentiment-analysis-dataset
[2] https://www.kaggle.com/datasets/ironbatshashank/brexit-twitter-data
[3] https://www.kaggle.com/code/lukaslonnroth/brexit-tweets-sentiment-analysis
[4] https://www.kaggle.com/datasets/chadjinik/labelledbrexittweets
[5] https://www.kaggle.com/datasets/kazanova/sentiment140
[6] https://www.kaggle.com/datasets/mmmarchetti/tweets-dataset
[7] https://www.kaggle.com/code/mdraselsarker/sentiment140-dataset-with-tweets
[8] https://www.kaggle.com/datasets/chadjinik/labelledbrexittweets/download
[9] https://www.kaggle.com/datasets/shashank1558/preprocessed-twitter-tweets

Selecciona el enlace (por número) donde quieres descargar los datos: 2

Seleccionaste el enlace: https://www.kaggle.com/datasets/ironbatshashank/brexit-twitter

Como se puede ver, a pesar de haber seleccionado el data set de brexit, terminamos por explorar el resto del conjunto de datos que se adaptaban con los criterios iniciales. Nuestra inclinacion al conjunto de datos [Twitter datasets](https://www.kaggle.com/datasets/mmmarchetti/tweets-dataset) fue luego de la exploracion de los diferentes conjuntos de datos, que se adaptaban con los criterios inciales de busqueda. Asi, descargamos este conjunto de datos, para ser almacendo en la nube de Google Colab.

### **Cuarta Tarea: Almacenamiento de datos**

  Tras la exitosa búsqueda y exploración de los conjuntos de datos de interés, particularmente aquellos localizados en el repositorio Kaggle, el siguiente paso crucial fue asegurar su almacenamiento adecuado para el procesamiento. Este almacenamiento se llevó a cabo mediante la carga directa del archivo descargado en el entorno de Google Colab.

  La decisión de almacenar los datos directamente en Colab nos permitio mejorar el rendimiento operativo de nuestras actividades de programacion.

  Nos facilito el acceso rápido y directo al conjunto de datos desde cualquier celda del notebook, eliminando la necesidad de descargas repetidas o configuraciones complejas cada vez que se inicia una sesión.

  I Al cargar el archivo directamente, se integramos el entorno de trabajo con el sistema de archivos temporal de Colab, facilitando su lectura y manipulación con librerías como Pandas, NumPy, etc.
  
  Una vez cargado, el conjunto de datos queda inmediatamente disponible en memoria o en el disco virtual de la sesión, listo para las fases posteriores de exploración, tratamiento y análisis, optimizando así el flujo de trabajo del proyecto.

  Por esta razon, luego del almacenamiento en Google colab, aplicamos un programa de carga y almacenamiento de datos en nuestro entorno de trabajo.

In [ ]:
import os

print("Archivos en el directorio actual:")
print(os.listdir())  # Esto te muestra qué archivos hay


Archivos en el directorio actual:
['.config', 'archive (7).zip', 'antropic.csv', 'tweets.csv', '.ipynb_checkpoints', 'sample_data', 'respuestas_agente_eda.csv', 'resultados_datasets.json']


In [ ]:
# prompt: tengo un archivo zip y quiero extraer el archivo csv contenido en el
import os
import re
import pandas as pd
import zipfile

# Especifica el nombre del archivo zip
zip_file_path = '/content/archive (7).zip'

# Especifica el nombre del archivo CSV dentro del zip (si lo conoces, de lo contrario tendrás que listarlos primero)
csv_file_name_in_zip = 'tweets.csv' # Cambia 'nombre_del_csv.csv' por el nombre real del archivo CSV

# Especifica la carpeta donde quieres extraer el archivo CSV
extract_path = '.' # Extraer en la carpeta actual

# Abre el archivo zip
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    # Lista los archivos dentro del zip (opcional, útil si no sabes el nombre exacto del CSV)
    print("Archivos dentro del ZIP:")
    zip_contents = zip_ref.namelist()
    print(zip_contents)

    # Extrae solo el archivo CSV específico
    if csv_file_name_in_zip in zip_contents:
        print(f"\nExtrayendo {csv_file_name_in_zip}...")
        zip_ref.extract(csv_file_name_in_zip, extract_path)
        print(f"Archivo extraído a: {os.path.join(extract_path, csv_file_name_in_zip)}")
    else:
        print(f"\nEl archivo '{csv_file_name_in_zip}' no se encontró dentro del ZIP.")

# Ahora puedes leer el archivo CSV extraído
# El archivo extraído estará en la ruta especificada por extract_path + csv_file_name_in_zip
# Por ejemplo, si extract_path es '.' (carpeta actual), la ruta sería 'nombre_del_csv.csv'
extracted_csv_path = os.path.join(extract_path, csv_file_name_in_zip)

try:
    df_csv = pd.read_csv(extracted_csv_path)
    print("\nContenido del archivo CSV extraído:")
    print(df_csv.head())
except FileNotFoundError:
    print(f"No se pudo encontrar el archivo extraído en: {extracted_csv_path}")
except Exception as e:
    print(f"Ocurrió un error al leer el archivo CSV: {e}")

Archivos dentro del ZIP:
['tweets.csv']

Extrayendo tweets.csv...
Archivo extraído a: ./tweets.csv

Contenido del archivo CSV extraído:
      author  \
0  katyperry   
1  katyperry   
2  katyperry   
3  katyperry   
4  katyperry   

                                                                                                                content  \
0                                             Is history repeating itself...?#DONTNORMALIZEHATE https://t.co/ngG11quhmK   
1  @barackobama Thank you for your incredible grace in leadership and for being an exceptional… https://t.co/ZuQLZpt6df   
2                                                                                   Life goals. https://t.co/XIn1qKMKQl   
3                                                                               Me right now 🙏🏻 https://t.co/gW55C1wrwd   
4                                                   SISTERS ARE DOIN' IT FOR THEMSELVES! 🙌🏻💪🏻❤️ https://t.co/0shuUYUBEv   

  country         date_time 

## **Exploracion de datos (EDA)**

  En el presente bloque se realizara una exploracion de datos del data frame seleccionado para el analisis sociologico y de marketing. Coherentemente con la propuesta desarrollada en la [metodologia](Metodologia) hemos elaborado un modelo de inteligencia artificial, buscando configurar un agente para la exploracion de datos. Al menos la siguiente tarea se enfocara en concretar los aspectos centrales de la exploracion de datos.

  Para llevar a cabo esta exploracion de datos configuramos un agente de Inteligencia Artificial utilizando **`"Langchain"`** y el modelo **`"llama3-70b-8192"`**  de **`"Groq"`**  para realizar un Análisis Exploratorio de Datos (EDA) interactivo sobre un DataFrame de Pandas. En esencia, permite hacer preguntas en lenguaje natural sobre tu dataset y obtener respuestas generadas por la IA, que a su vez ejecuta código Python para obtener la información.

  En otras palabras, haciendo uso de las capacidades de procesamiento de **`"Groq"`** para la interaccion de nuestro entorno y codigo python con el modelo Llms **`"llama3-70b-8192"`** alojado. Definimos las solicitudes de calculos estadisticos mediante el manejo y la generacion de codigo python para el procesamiento de la informacion de nuestro data frame. Las respuestas de nuestros prompts haran disponible de forma legible nuestras peticiones.

  De este modo, agregamos un paso adicional intermedio entre la configuracion de la plantilla de instrucciones y la configuracion de la cadena rag. El paso fue la creacion y especificacion del agente.
  
  Para ello realizamos primero la siguiente plantilla, indicando las necesidades en la exploracion de datos.

    """ Eres un experto analista de datos que responde preguntas
    sobre un DataFrame `df` usando pandas.

    Tu objetivo es:
    Responder en español.
    No inventar datos ni columnas.
    Usar pandas internamente para los cálculos.
    Decir si una columna o dato no existe.
    Realizar análisis exploratorio simple (EDA).

>
    Ejemplos:
    Pregunta: ¿Cuántas filas hay?
    Respuesta: El DataFrame tiene 1432 filas.
    Pregunta: ¿Qué columnas tienen valores nulos?
    Respuesta: Las columnas 'edad' y 'salario' tienen valores nulos.
    Pregunta: ¿Cuáles son las columnas categóricas?
    Respuesta: Son aquellas con tipo 'object' o que tienen pocos valores únicos,   como 'ciudad' y 'sexo'.
    Ahora responde la siguiente pregunta: {input}"""

En este caso, como Pandas no es una herramienta externa, sino una herramienta que el agente de IA utiliza y controla directamente, se incluyen las instrucciones para que el agente genere código. El agente sera capaz de usar la herramienta (Pandas) como parte de su propio proceso de razonamiento y respuesta.

Por ello, posteriormente, a la creacion de la plantilla de instrucciones, procedimos a la creacion del Agente de Anallisis Exploratorio. La creación del Agente de Análisis Exploratorio se enfoco en definir una funcion central, que pudiera acceder al data frame para interactuar con el modelo de lenguaje **`"llama3-70b-8192"`** (instanciado como llm) definido previamente. Como se observa en el siguiente fragmento del codigo:

>
    agent = create_pandas_dataframe_agent(
    llm=llm,
    df=df,
    prompt=prompt,
    verbose=True,
    allow_dangerous_code=True,
    handle_parsing_errors=True # <-- Add this argument)

Posteriormente, con el agente ya creado y configurado, el siguiente paso fue especificar las solicitudes (preguntas) de análisis que se le harían. Se configuró una lista de preguntas directas que el agente debía responder basándose en el dataframe al que tenía acceso.

Para cada una de estas preguntas, el agente interpretaría la solicitud en lenguaje natural, generaría internamente el código Python/Pandas necesario para obtener la respuesta, lo ejecutaría sobre el df, y luego formularía la respuesta final en español.

>
    "¿Cuántas filas y columnas tiene el dataset?",
    "¿Cuáles son los nombres de las columnas y su tipo estructura?",
    "¿Qué columnas tienen valores nulos?",
    "¿Cuál es el porcentaje de valores nulos por columna?",
    "¿Hay valores duplicados en el dataset?",
    "¿Cuáles son los valores únicos de las columnas categoricas",
    "¿Cual es el porcentaje de los valores de las columnas bool?",
    "¿Cuál es el valor máximo y mínimo de la columnas numericas ?",
    "¿Cuál es la media, mediana y desviación estándar de cada columna numérica ?"
Cada una de estas preguntas fue procesada secuencialmente por el agente, y sus respuestas fueron capturadas para un análisis posterior, demostrando la capacidad del sistema para automatizar y agilizar la fase de exploración de datos de manera interactiva.

Asi que luego de cargar el conjunto de datos, realizamos este analisis exploratorio con el modelo de lenguaje generativo, incluyendo la incorporacion herramientas PANDAS para que el agente interactue con el data frame, segun nuestras solicitudes.

#### Carga del conjunto de datos

In [ ]:
df_csv.head()

author  \
0  katyperry   
1  katyperry   
2  katyperry   
3  katyperry   
4  katyperry   

                                                                                                                content  \
0                                             Is history repeating itself...?#DONTNORMALIZEHATE https://t.co/ngG11quhmK   
1  @barackobama Thank you for your incredible grace in leadership and for being an exceptional… https://t.co/ZuQLZpt6df   
2                                                                                   Life goals. https://t.co/XIn1qKMKQl   
3                                                                               Me right now 🙏🏻 https://t.co/gW55C1wrwd   
4                                                   SISTERS ARE DOIN' IT FOR THEMSELVES! 🙌🏻💪🏻❤️ https://t.co/0shuUYUBEv   

  country         date_time            id language  latitude  longitude  \
0     NaN  12/01/2017 19:52  8.196330e+17       en       NaN        NaN   
1     NaN  11/01/2017 08:38  8.191010e+17       en       NaN        NaN   
2     NaN  11/01/2017 02:52  8.190140e+17       en       NaN        NaN   
3     NaN  11/01/2017 02:44  8.190120e+17       en       NaN        NaN   
4     NaN  10/01/2017 05:22  8.186890e+17       en       NaN        NaN   

   number_of_likes  number_of_shares  
0             7900              3472  
1             3689              1380  
2            10341              2387  
3            10774              2458  
4            17620              4655

### Configuracion de nuestro agente:
- Instalacion de las librerias necesarias
- Configuracion de la API KEY
- Creacion de las plantillas de instrucciones
- Configuracion del agente para la interaccion con el data frame mediante el uso de PANDAS

In [ ]:
!pip install -q transformers
!pip install -q transformers ipywidgets

In [ ]:
# Install chromadb
!pip install -q chromadb

In [ ]:
# 📦 1. Instala dependencias necesarias
!pip install langchain pandas langchain-groq langchain-experimental -q

In [ ]:
# 📚 2. Importa librerías
import pandas as pd
import os
from langchain_groq import ChatGroq
# Update the import path for create_pandas_dataframe_agent
from langchain_experimental.agents import create_pandas_dataframe_agent
from langchain.prompts import PromptTemplate

# 🔐 3. Configura tu clave API de Groq
os.environ["GROQ_API_KEY"] = "GROQ_API_KEY"  # <-- Reemplaza por tu clave real

# 📂 4. Carga tu dataset CSV
# The correct path seems to be the one extracted earlier
df = pd.read_csv("/content/tweets.csv")  # <-- Use the correct file path

# 🤖 5. Carga el modelo LLaMA 3 desde Groq
llm = ChatGroq(
    groq_api_key=os.environ["GROQ_API_KEY"],
    model_name="llama3-70b-8192"
)

# 🧠 6. Crea un prompt personalizado para análisis de datos
prompt = PromptTemplate.from_template("""
Eres un experto analista de datos que responde preguntas sobre un DataFrame `df` usando pandas.

Tu objetivo es:
- Responder en español.
- No inventar datos ni columnas.
- Usar pandas internamente para los cálculos.
- Decir si una columna o dato no existe.
- Realizar análisis exploratorio simple (EDA).

Ejemplos:
Pregunta: ¿Cuántas filas hay?
Respuesta: El DataFrame tiene 1432 filas.

Pregunta: ¿Qué columnas tienen valores nulos?
Respuesta: Las columnas 'edad' y 'salario' tienen valores nulos.

Pregunta: ¿Cuáles son las columnas categóricas?
Respuesta: Son aquellas con tipo 'object' o que tienen pocos valores únicos, como 'ciudad' y 'sexo'.

Ahora responde la siguiente pregunta:
{input}
""")

# 🧠 7. Crea el agente de análisis exploratorio con pandas
agent = create_pandas_dataframe_agent(
    llm=llm,
    df=df,
    prompt=prompt,
    verbose=True,
    allow_dangerous_code=True,
    handle_parsing_errors=True # <-- Add this argument
)

/usr/local/lib/python3.11/dist-packages/langchain_experimental/agents/agent_toolkits/pandas/base.py:283: UserWarning: Received additional kwargs {'prompt': PromptTemplate(input_variables=['input'], input_types={}, partial_variables={}, template="\nEres un experto analista de datos que responde preguntas sobre un DataFrame `df` usando pandas.\n\nTu objetivo es:\n- Responder en español.\n- No inventar datos ni columnas.\n- Usar pandas internamente para los cálculos.\n- Decir si una columna o dato no existe.\n- Realizar análisis exploratorio simple (EDA).\n\nEjemplos:\nPregunta: ¿Cuántas filas hay?\nRespuesta: El DataFrame tiene 1432 filas.\n\nPregunta: ¿Qué columnas tienen valores nulos?\nRespuesta: Las columnas 'edad' y 'salario' tienen valores nulos.\n\nPregunta: ¿Cuáles son las columnas categóricas?\nRespuesta: Son aquellas con tipo 'object' o que tienen pocos valores únicos, como 'ciudad' y 'sexo'.\n\nAhora responde la siguiente pregunta:\n{input}\n"), 'handle_parsing_errors': True} 

- Configuracion de las solicitudes especificas del agente
- Impresion de los resultados
- Resguardo de los resultados en un documento JSON

In [ ]:
# 🧾 Lista para guardar resultados
resultados = []

# ❓ 8. Ejecuta preguntas y guarda respuestas
preguntas = [
    "¿Cuántas filas y columnas tiene el dataset?",
    "¿Cuáles son los nombres de las columnas y su tipo estructura?",
    "¿Qué columnas tienen valores nulos?",
    "¿Cuál es el porcentaje de valores nulos por columna?",
    "¿Hay valores duplicados en el dataset?",
    "¿Cuáles son los valores únicos de las columnas categoricas",
    "¿Cual es el porcentaje de los valores de las columnas bool?",
    "¿Cuál es el valor máximo y mínimo de la columnas numericas ?",
    "¿Cuál es la media, mediana y desviación estándar de cada columna numérica ?"
]

for pregunta in preguntas:
    print("🔍 Pregunta:", pregunta)
    try:
        respuesta = agent.run(pregunta)
        print("🧠 Respuesta:", respuesta)
        print("=" * 100)

        # Guarda en lista
        resultados.append({
            "pregunta": pregunta,
            "respuesta": respuesta,
            "modelo": llm.model_name,# cambiar segun el modelo usado
            "archivo_analizado": "president_county_candidate.csv"# cambiar
        })
    except Exception as e: # Catching a general Exception for robustness
        print(f"❌ Error processing question '{pregunta}': {e}")
        print("=" * 100)
        # Optionally, store the error details instead of a successful response
        resultados.append({
            "pregunta": pregunta,
            "respuesta": f"Error: {e}",
            "modelo": llm.model_name,
            "archivo_analizado": "president_county_candidate.csv"
        })


# 🗃️ Convertir a DataFrame
df_resultados = pd.DataFrame(resultados)

# 💾 Guardar como CSV
df_resultados.to_csv("/content/respuestas_agente_eda.csv", index=False)

🔍 Pregunta: ¿Cuántas filas y columnas tiene el dataset?


> Entering new AgentExecutor chain...
Thought: To find the number of rows and columns in the dataset, I can use the `shape` attribute of the dataframe.

Action: python_repl_ast
Action Input: df.shape(52542, 10)I now know the final answer
Final Answer: El dataset tiene 52542 filas y 10 columnas.

> Finished chain.
🧠 Respuesta: El dataset tiene 52542 filas y 10 columnas.
🔍 Pregunta: ¿Cuáles son los nombres de las columnas y su tipo estructura?


> Entering new AgentExecutor chain...
Thought: I need to check the column names and their data types.

Action: python_repl_ast
Action Input: df.info()<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52542 entries, 0 to 52541
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   author            52542 non-null  object 
 1   content           52542 non-null  object 
 2   country           36 non-null     object 


#### Impresion de los resultados del Analisis Exploratorio

In [ ]:
# Muestra todo el contenido de cada columna (sin cortes)
pd.set_option('display.max_colwidth', None)

# Muestra todas las filas (¡cuidado si son muchas!)
pd.set_option('display.max_rows', None)

# Muestra todas las columnas si tuvieras más de dos
pd.set_option('display.max_columns', None)

# Ahora imprime
print(df_resultados)

                                                                      pregunta  \
0                                  ¿Cuántas filas y columnas tiene el dataset?   
1                ¿Cuáles son los nombres de las columnas y su tipo estructura?   
2                                          ¿Qué columnas tienen valores nulos?   
3                         ¿Cuál es el porcentaje de valores nulos por columna?   
4                                       ¿Hay valores duplicados en el dataset?   
5                   ¿Cuáles son los valores únicos de las columnas categoricas   
6                  ¿Cual es el porcentaje de los valores de las columnas bool?   
7                 ¿Cuál es el valor máximo y mínimo de la columnas numericas ?   
8  ¿Cuál es la media, mediana y desviación estándar de cada columna numérica ?   

                                                                                                                                                                                   

### Resultados del Analisis Exploratorio
---
Luego de un análisis exploratorio inicial de nuestro dataset se comprende mejor la estructura y la calidad de sus datos. Entre los resultados solicitados obtuvimos la sigiente informacion:

> En resumen el dataset contiene 52542 filas y 10 columnas.

Esta informacion nos permite corroborar la extension de nuestra fuente de datos para conseguir analisis e interpretaciones ajustadas a grupos representativos de la opinion publica en Twitter.

> Los puntos clave del análisis incluye la identificación de las columnas y sus respectivos tipos de datos, como se muestra en la figura anterior que indica la siguiente información:

Con respecto al tipo de objeto de las variables nos percatamos que el conjunto de datos tenia objetos de string, numericos y decimales. De hecho, las variables de tipo object, lo que generalmente indica datos de texto o strings son:

    'author',
    'content',
    'country',
    'date_time',
    'language':

Mientras que las variables de tipo float64, nos sugiere que las categorias a continuacion son números de punto flotante:

    'id',
    'latitude',
    'longitude'
Por ultimo, nuestras variables de tipo int64 indican que las siguientes categorias son de indicando números enteros

    'number_of_likes',
    'number_of_shares'

Por otro lado, la inspección de los valores nulos nos arrojó que las columnas  'country', 'latitude' y 'longitude' contienen valores nulos, a diferencia del resto de las columnas no tienen valores nulos. De hecho, según el  porcentaje de valores nulos para cada columna, nos muestra que las variables decimales tienen una alta proporción de datos faltantes, como indicamos:

    'country': Aproximadamente el 99.93% de sus valores son nulos.
    'latitude': Aproximadamente el 99.998% de sus valores son nulos.
    'longitude': Aproximadamente el 99.998% de sus valores son nulos.
    Las demás columnas tienen 0% de valores nulos.

Esta exploracion es alentadora para nuestros propositos, porque la colummna content no contiene ningun valor nulo, asi como las columnas sobre el numero shares o likes de las publicaciones.

>Por otro lado determinamos que **no hay valores duplicados** en el dataset, lo cual es un buen indicio de que la combinación de las variables se refieren a un registro único, y, por tanto de la homogeneidad de la estructura de la base de datos.

Otra inspección importante se refiere a la cardinalidad de los valores únicos dentro de las columnas identificadas como categóricas (tipo object). En esta inspección cercioramos que los valores de las categorías responden a una amplitud de valores como los que resumimos a continuación por variable:

    'author': Contiene una lista limitada de autores de los tweets,
    incluyendo 'katyperry', 'justinbieber', 'BarackObama', etc.
    'content': Presenta una gran cantidad de valores únicos como
    registros de contenido de los twitters.
    'country': Muchos valores únicos.
    'date_time': este campo igualmente ofrece muchos valores únicos por los formatos
    de la información que se da por minuto, hora, día, semana, mes, años.
    'language': Muestra una variedad de idiomas, incluyendo 'en' (inglés),
    'es' (español), 'pt' (portugués), etc.

Otro hallazgo de la exploracion de datos es que, no se encontraron columnas de tipo booleano en el dataset.

En cuanto a las estadísticas descriptivas básicas podemos mencionar aquellas de las que vamos hacer uso: el número de likes y el número de  tweets compartidos.
1.  **`"number_of_likes"`**:
- Media: 9637.84
- Mediana: 2595.50
- Desviación estándar: 18759.08
- Mínimo: 0
- Máximo: 429159

2. **`"number_of_shares"`**:
- Media: 5386.88
- Mediana: 1266.00
- Desviación estándar: 11517.26
- Mínimo: 0
- Máximo: 219062

Las diferencias entre la media y la mediana para 'number_of_likes' y 'number_of_shares' sugieren que estas columnas pueden tener una distribución asimétrica. Es decir, algunos pocos registros con muchos likes y shares; mientras que muchos otros sin ninguno de los valores. En efecto, esta apreciacion coincide con la gran desviacion estandar de nuestros datos, que nos revelan la presencia de una posible distribucion en ley de potencia. Esta observacion conjuntamente con valores minimos de 0 likes y shares, y valores maximos altos de ambas variables insinuan precisamente que, hay unas publicaciones con muchisimos likes y shares, y otras muchas publicaciones con pocas.




## **Modelado y extraccion de temas**


En este punto, luego de la exploracion de los datos, y confirmar la validez estadissticas de los mismos, y sus especificaciones, se hace posible la ejecucion de nuestro procesamiento de datos. Para ello, es necesario la creacion de un cliente que permita gestionar la conexión y la interacción con el servicio de IA. El servicio requerido es un patrón común y necesario en el desarrollo con APIs de LLMs. Es como tener un control remoto (el objeto cliente) para operar un dispositivo (el modelo de IA en la nube).

En este caso, configuramos el modelo **`"claude-3-haiku-20240307"`** de **`"Antropic"`**. Una vez configurado el modelo, crea el objeto especial llamado cliente. Este cliente servira como interfaz directa para interactuar con los servidores de Anthropic. Este objeto ya tiene toda la lógica para enviar tus solicitudes y recibir las respuestas del modelo.

Para ello, configuramos la función *analyze_text_con_claude*, un componente crucial para el procesamiento inicial y la extracción de información clave de tus textos. Esta función combina la precisión de las expresiones regulares con la capacidad de comprensión avanzada de un Modelo de Lenguaje Grande (LLM) para automatizar el análisis de contenido.

>Analiza texto para extraer hashtags, URLs, y los 3 temas principales, usando Anthropic Claude para temas y regex para lo demás.

>Args:
>   content (str): El texto que quieres analizar.

> Returns:

> dict: Un diccionario con "hashtags", "urls", y "temas".

> Incluye mensajes de error si hay problemas con la API, el formato,
> o si el contenido de entrada es inválido.

Luego, la función *analyze_text_con_claude* está diseñada para la extraccion de entidades específicas, utilizando expresiones regulares (re) con lo cual se identifico y extrajo patrones simples pero importantes (Hashtags y URLs).

La tarea más compleja, la extracción de los 3 temas más relevantes del texto, se delega al modelo Anthropic Claude (específicamente claude-3-haiku-20240307).

Instrucciones al LLM (el Prompt): La función construye un prompt detallado que se envía directamente a Claude. El prompt construido fue el siguiente:

> You are an expert text analysis assistant.
  Analyze the following text and return the 3 most relevant main themes or keywords.
  The result must be a valid JSON object with a "themes" key containing a list of strings.
  Ensure the output is *only* the JSON object, with no additional text before or after.

    JSON output example:
    {{
      "themes": ["keyword 1", "keyword 2", "keyword 3"]
      }}
    Text to analyze:"""

Este prompt le indica a Claude que actúe como un "asistente experto en análisis de texto", que analice el contenido provisto y que devuelva los 3 temas principales en un formato JSON estandarizado (una lista bajo la clave "themes").

Con este prompt, pudimos configurar la interacción con la API. Utilizamos el objeto client (previamente configurado con tu API Key de Anthropic) para enviar el prompt al modelo Claude en la nube y recibir su respuesta.

Esta conexion optimmiza el procesamiento de la Respuesta. Una vez que Claude devuelve su respuesta (esperando que sea un JSON válido), la función la parsea para extraer los temas. Se incluye un robusto manejo de errores para lidiar con textos de entrada inválidos, problemas de comunicación con la API de Claude o si Claude no devuelve el JSON en el formato esperado, se asegura que la función no falle inesperadamente.

Finalmente, la función consolida toda la información extraída (hashtags, URLs y temas) en un único diccionario, facilitando su uso en análisis posteriores.

Es importante destacar que el prompt utilizado dentro de esta función es específico para instruir a Claude sobre la tarea de extracción de temas. Esto es distinto de la plantilla de instrucciones general que se utiliza para configurar el agente de LangChain que opera sobre DataFrames. Cada plantilla cumple un rol particular para el "actor" de IA al que va dirigida. En otras palabras, definimos una plantilla de instrucciones (el prompt_instructions) dentro de tu función analyze_text_con_claude que está específicamente diseñada para guiar a Claude en la tarea de extracción de los tres temas principales y asegurar que la respuesta se registre en la estructura de diccionario (JSON) que esperas.

### Configuracion del modelo Antropic

In [ ]:
# Install chromadb
!pip install -q chromadb

In [ ]:
!pip install langchain-anthropic

In [ ]:
## Configuracion del agente

import os
import re
import json
import pandas as pd
from anthropic import Anthropic


# --- 1.1. Configure Your Anthropic API Key ---

os.environ["ANTHROPIC_API_KEY"] = "ANTROPIC_API_KEY"
ANTHROPIC_API_KEY = os.environ.get("ANTHROPIC_API_KEY")


if not ANTHROPIC_API_KEY:
    raise ValueError(
        "Error! The 'ANTHROPIC_API_KEY' environment variable isn't set. "
        "Please configure it before running the code."
    )

# --- 1.2. Initialize the Anthropic Client ---
client = Anthropic(api_key=ANTHROPIC_API_KEY)

# --- 1.3. Select the Claude Model ---
ANTHROPIC_MODEL = "claude-3-haiku-20240307"

print(f"Setup complete. Using Anthropic model: {ANTHROPIC_MODEL}")

Setup complete. Using Anthropic model: claude-3-haiku-20240307


### Creacion de la funcion basada en una interaccion conversacional con el modelo.

In [ ]:
def analyze_text_con_claude(content: str) -> dict:
    """
    Analiza texto para extraer hashtags, URLs, y los 3 temas principales,
    usando Anthropic Claude para temas y regex para lo demás.

    Args:
        content (str): El texto que quieres analizar.

    Returns:
        dict: Un diccionario con "hashtags", "urls", y "temas".
              Incluye mensajes de error si hay problemas con la API, el formato,
              o si el contenido de entrada es inválido.
    """
    # --- ¡NUEVO! 2.1. Validación del Contenido de Entrada ---
    # Comprobamos si el contenido no es una cadena o si está vacío/solo espacios.
    if not isinstance(content, str) or not content.strip():
        # Devolvemos un diccionario de error explícito en lugar de dejar que la función falle.
        return {
            "hashtags": [],
            "urls": [],
            "temas": [f"Error: Contenido de texto inválido o vacío (tipo: {type(content).__name__})."]
        }

    # --- 2.2. Extraer Hashtags (con expresiones regulares) ---
    hashtags = list(set(re.findall(r'#(\w+)', content)))

    # --- 2.3. Extraer URLs (con expresiones regulares) ---
    url_pattern = re.compile(
        r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
    )
    urls = url_pattern.findall(content)

    # --- 2.4. Identificar Temas Principales (con Anthropic Claude) ---
    extracted_themes = []

    # Preparamos el "prompt" o instrucción para Claude.
    prompt_instructions = f"""
    You are an expert text analysis assistant.
    Analyze the following text and return the 3 most relevant main themes or keywords.
    The result must be a valid JSON object with a "themes" key containing a list of strings.
    Ensure the output is *only* the JSON object, with no additional text before or after.

    JSON output example:
    {{
      "themes": ["keyword 1", "keyword 2", "keyword 3"]
    }}

    Text to analyze:
    {content}
    """

    # Preparamos una versión truncada del contenido para los mensajes de error,
    # para que no imprimamos textos muy largos en la consola.
    truncated_content_for_error = content[:75] + "..." if len(content) > 75 else content


    try:
        api_response = client.messages.create(
            model=ANTHROPIC_MODEL,
            max_tokens=200,
            messages=[
                {"role": "user", "content": prompt_instructions}
            ]
        )

        response_text = api_response.content[0].text.strip()
        json_response = json.loads(response_text)

        if "themes" in json_response and isinstance(json_response["themes"], list):
            extracted_themes = json_response["themes"][:3]
        else:
            print(f"Warning: Unexpected theme format for text: {truncated_content_for_error}. Raw response: {response_text}")
            extracted_themes = [response_text]

    except json.JSONDecodeError as e:
        print(f"Warning: Claude didn't return valid JSON for text: {truncated_content_for_error}. Error: {e}")
        print(f"Raw Claude response: '{response_text}'")
        extracted_themes = ["Error: Invalid JSON from Claude.", response_text]
    except Exception as e:
        print(f"Warning: General error calling Claude API for text: {truncated_content_for_error}. Error: {e}")
        extracted_themes = ["Error extracting themes with Claude."]

    # --- 2.5. Devolver todos los resultados ---
    return {
        "hashtags": hashtags,
        "urls": urls,
        "temas": extracted_themes
    }

### Aplicacion del modelo en el data set muestreado.

Debido a la magnitud de registo de nuestro data frame, hemos decidido hacer un muestreo de la informacion del 10%. La muestra del 10% de los registros permitiran computar y procesar la informacion adecuada para nuestros objetivos de estudio.

De manera que, la aplicacion del analisis de texto orientado por nuestras solicitudes extraera los temas, hashtags y url's, y posteriormente, identificara los temas principales de cada registro. El resultado sera una columna adicional con la lista de url's, temas y hashtags asociado a cada texto publicado.

In [ ]:
# --- 3.2. Configuración del Muestreo (Opcional, pero recomendado para pruebas) ---
# Define cuántas filas quieres muestrear para probar.
# Si lo pones a 0, procesará todo el DataFrame.
NUM_ROWS_TO_SAMPLE = int(len(df_csv) * 0.1) # <--- ¡CAMBIA ESTE NÚMERO SEGÚN TUS NECESIDADES DE PRUEBA!

# --- 3.3. Realiza el Muestreo del DataFrame (si se configuró) ---
df_to_process = None
if 'content' not in df_csv.columns:
    print("\n¡Error! El DataFrame 'df_csv' no tiene la columna 'content'. Asegúrate de que tus textos estén en esa columna.")
elif NUM_ROWS_TO_SAMPLE > 0 and NUM_ROWS_TO_SAMPLE <= len(df_csv):
    df_to_process = df_csv.sample(n=NUM_ROWS_TO_SAMPLE, random_state=42)
    print(f"\n--- 3.3. Se ha creado una muestra de {len(df_to_process)} filas de 'df_csv' para el análisis ---")
    print(df_to_process[['content']])
else:
    print("\n--- 3.3. No se ha configurado un muestreo válido o el número excede el tamaño del DF. Se procesará 'df_csv' completo ---")
    df_to_process = df_csv

# --- 3.4. Aplicar la Función de Análisis a la MUESTRA (o al DF completo) ---
if df_to_process is not None and not df_to_process.empty:
    print(f"\n--- 3.4. Iniciando el análisis de texto (usando {ANTHROPIC_MODEL}). Esto puede tardar y generar costos de API ---")

    # Aplicamos la función 'analyze_text_con_claude' a cada texto de la columna 'content'.
    df_to_process['analysis_results'] = df_to_process['content'].apply(analyze_text_con_claude)

    print("\n--- 3.4. Análisis completado en la muestra (o DataFrame completo) ---")
    print("Primeras filas con los resultados del análisis:")
    print(df_to_process[['content', 'analysis_results']].to_string())

    print("\n--- Detalles del análisis de los primeros 5 textos analizados (ejemplo): ---")
    if not df_to_process.empty:
        for i in range(min(5, len(df_to_process))):
            print(f"\nResultado para la fila {df_to_process.index[i]}:")
            print(df_to_process['analysis_results'].iloc[i])

else:
    print("\n¡Error! El DataFrame no pudo ser preparado para el análisis. Revisa los mensajes anteriores.")

Se truncaron las últimas líneas 5000 del resultado de transmisión.
35080                                            @mirtha_michelle book launch din! Letters, To the Men I Have Loved. Congrats mama! Besos x http://t.co/Vx6AjFBLS6                                                                                                              {'hashtags': [], 'urls': ['http://t.co/Vx6AjFBLS6'], 'temas': ['book launch', 'letters', 'love']}
38842                                Wing part found on Tanzanian island "highly likely" from #MH370, Australia's transport minister says. https://t.co/k7JLJ7iDN7                                                                                           {'hashtags': ['MH370'], 'urls': ['https://t.co/k7JLJ7iDN7'], 'temas': ['MH370', 'aircraft debris', 'investigation']}
34506                                                                                      @taylordunn73 literally had a brownie and 5 friends. Best bday, EVER ☺️                               

____
Resultado

El resultado de la ejecucion del codigo es la incorporacion de una variable analizada con la extraccion de los urls, hashgtags y los tres temas principales de cada uno de los registros de nuestro data frame. Esto nos permitira profundizar el analisis e interpretacion de los temas, y su influencia, ya que, al conservar los registros en cada una de las columnas, podremos vincularlos con los numeros de likes o hashtags que obtienen.


### Post-procesamiento del modelado de temas

Después de la fase de extracción y análisis de texto, el dataframe se enriqueció con un diccionario en cada registro, conteniendo los temas, hashtags y URLs identificados. Para avanzar en los objetivos de estudio, era crucial realizar un paso de post-procesamiento que, a su vez, actuaría como un preprocesamiento vital para el análisis e interpretación subsiguiente.

Esta etapa consistió en dos acciones principales:

- Confirmación y Validación de Datos:
Se llevó a cabo una revisión de los datos registrados por el modelo de análisis y extracción de temas. Esto asegura que la información (temas, hashtags, URLs) fuera precisa y estuviera en el formato esperado, sirviendo como un control de calidad inicial.
- Expansión de Variables: Para facilitar el análisis posterior, las listas de temas, hashtags y URLs contenidas en el diccionario de cada registro se expandieron en variables separadas dentro del dataframe. Este proceso transforma los datos anidados en columnas individuales y accesibles, lo que simplifica enormemente las operaciones de filtrado, agrupación y visualización en las etapas de análisis e interpretación.

Este post-procesamiento es fundamental para transformar los resultados crudos del modelo en un formato utilizable y estructurado, maximizando el valor de la información extraída.

A su vez, es importante mencionar que los resultados se guardaron en un documento csv, para facilitar los procesammientos posteriores, sin necesidad de iniciar todo el proceso nuevamente.

#### Inspeccion de las variables resultantes

In [ ]:
df_to_process.columns

Index(['author', 'content', 'country', 'date_time', 'id', 'language',
       'latitude', 'longitude', 'number_of_likes', 'number_of_shares',
       'analysis_results'],
      dtype='object')

#### Exportacion a CSV del conjunto de datos enriquecido



In [ ]:
# Change the file path to a valid directory on the Linux system
# For example, save it in the /content/ directory or the current directory ('.')
df_to_process.to_csv('/content/antropic.csv', index=False)

### Importacion del conjunto de datos enriquecido para la expension y posterior analisis

In [ ]:
df=pd.read_csv('/content/antropic.csv')

In [ ]:
df.head()

author  \
0  KimKardashian   
1  britneyspears   
2      Cristiano   
3   justinbieber   
4        Twitter   

                                                                                                                                        content  \
0                                                                               #TheBFFIssue @sorbetmag @riccardotisci17 http://t.co/VmVqbZ0OtZ   
1  Thank you to @mtv for premiering #SlumberParty on their official snapchat! I’ll be taking over their account tomorrow. Follow them!!! 👻: MTV   
2                                                                    What an exciting game at the Bernabéu! Hala Madrid! http://t.co/R2uUpf9o3x   
3                                                                                                           Mine too :) https://t.co/nTj4mytOf9   
4                             Looks like @FIFAcom is on Twitter just in time to spread big news about the World Cup directly from South Africa!   

  country         date_time            id language  latitude  longitude  \
0     NaN  26/09/2015 21:18  6.478830e+17      und       NaN        NaN   
1     NaN  18/11/2016 08:34  7.995310e+17       en       NaN        NaN   
2     NaN  23/10/2013 21:55  3.931340e+17       en       NaN        NaN   
3     NaN  07/11/2015 02:41  6.628220e+17       en       NaN        NaN   
4     NaN  09/06/2010 20:49  1.580143e+10       en       NaN        NaN   

   number_of_likes  number_of_shares  \
0             2891               700   
1             9121              3208   
2             8624             11800   
3            44867             31307   
4               56               297   

                                                                                                        analysis_results  
0  {'hashtags': ['TheBFFIssue'], 'urls': ['http://t.co/VmVqbZ0OtZ'], 'temas': ['friendship', 'social media', 'fashion']}  
1                 {'hashtags': ['SlumberParty'], 'urls': [], 'temas': ['television', 'social media', 'event promotion']}  
2                   {'hashtags': [], 'urls': ['http://t.co/R2uUpf9o3x'], 'temas': ['football', 'sports', 'Real Madrid']}  
3     {'hashtags': [], 'urls': ['https://t.co/nTj4mytOf9'], 'temas': ['social media', 'personal expression', 'brevity']}  
4                                           {'hashtags': [], 'urls': [], 'temas': ['World Cup', 'FIFA', 'South Africa']}

#### Pequeña inspeccion del numero de registros y el tipo de objetos

In [ ]:
len(df_to_process)

5254

In [ ]:
df_to_process.dtypes

author               object
content              object
country              object
date_time            object
id                  float64
language             object
latitude            float64
longitude           float64
number_of_likes       int64
number_of_shares      int64
analysis_results     object
dtype: object

In [ ]:
df_to_process.head()

author  \
28263  KimKardashian   
28996  britneyspears   
32910      Cristiano   
3834    justinbieber   
23410        Twitter   

                                                                                                                                            content  \
28263                                                                               #TheBFFIssue @sorbetmag @riccardotisci17 http://t.co/VmVqbZ0OtZ   
28996  Thank you to @mtv for premiering #SlumberParty on their official snapchat! I’ll be taking over their account tomorrow. Follow them!!! 👻: MTV   
32910                                                                    What an exciting game at the Bernabéu! Hala Madrid! http://t.co/R2uUpf9o3x   
3834                                                                                                            Mine too :) https://t.co/nTj4mytOf9   
23410                             Looks like @FIFAcom is on Twitter just in time to spread big news about the World Cup directly from South Africa!   

      country         date_time            id language  latitude  longitude  \
28263     NaN  26/09/2015 21:18  6.478830e+17      und       NaN        NaN   
28996     NaN  18/11/2016 08:34  7.995310e+17       en       NaN        NaN   
32910     NaN  23/10/2013 21:55  3.931340e+17       en       NaN        NaN   
3834      NaN  07/11/2015 02:41  6.628220e+17       en       NaN        NaN   
23410     NaN  09/06/2010 20:49  1.580143e+10       en       NaN        NaN   

       number_of_likes  number_of_shares  \
28263             2891               700   
28996             9121              3208   
32910             8624             11800   
3834             44867             31307   
23410               56               297   

                                                                                                            analysis_results  
28263  {'hashtags': ['TheBFFIssue'], 'urls': ['http://t.co/VmVqbZ0OtZ'], 'temas': ['friendship', 'social media', 'fashion']}  
28996                 {'hashtags': ['SlumberParty'], 'urls': [], 'temas': ['television', 'social media', 'event promotion']}  
32910                   {'hashtags': [], 'urls': ['http://t.co/R2uUpf9o3x'], 'temas': ['football', 'sports', 'Real Madrid']}  
3834      {'hashtags': [], 'urls': ['https://t.co/nTj4mytOf9'], 'temas': ['social media', 'personal expression', 'brevity']}  
23410                                           {'hashtags': [], 'urls': [], 'temas': ['World Cup', 'FIFA', 'South Africa']}

#### Expansion del conjunto de datos en variables de cada una de las listas -URLs, Hashtags y temas- del diccionario regitrado

In [ ]:
df_expandido = df_to_process['analysis_results'].apply(pd.Series)


#### Visualizacion de los registros expandidos y concatenacion con el data frame

En síntesis, ejecutaremos dos líneas de código para la transformacion de una columna de datos anidados (diccionarios) en múltiples columnas nuevas y directamente accesibles, lo que facilita enormemente cualquier análisis, filtrado o visualización posterior de los resultados de tu extracción de temas. Es un paso de normalización de datos esencial cuando trabajas con salidas de modelos de IA que a menudo vienen en formatos complejos.

In [ ]:
df_expandido.head()

hashtags                       urls  \
28263   [TheBFFIssue]   [http://t.co/VmVqbZ0OtZ]   
28996  [SlumberParty]                         []   
32910              []   [http://t.co/R2uUpf9o3x]   
3834               []  [https://t.co/nTj4mytOf9]   
23410              []                         []   

                                              temas  
28263           [friendship, social media, fashion]  
28996   [television, social media, event promotion]  
32910               [football, sports, Real Madrid]  
3834   [social media, personal expression, brevity]  
23410               [World Cup, FIFA, South Africa]

In [ ]:
df_expandido2 = df_to_process['analysis_results'].apply(pd.Series)
df_completo = pd.concat([df_to_process.drop(columns=['analysis_results']), df_expandido2], axis=1)


In [ ]:
df_completo.head()

author  \
28263  KimKardashian   
28996  britneyspears   
32910      Cristiano   
3834    justinbieber   
23410        Twitter   

                                                                                                                                            content  \
28263                                                                               #TheBFFIssue @sorbetmag @riccardotisci17 http://t.co/VmVqbZ0OtZ   
28996  Thank you to @mtv for premiering #SlumberParty on their official snapchat! I’ll be taking over their account tomorrow. Follow them!!! 👻: MTV   
32910                                                                    What an exciting game at the Bernabéu! Hala Madrid! http://t.co/R2uUpf9o3x   
3834                                                                                                            Mine too :) https://t.co/nTj4mytOf9   
23410                             Looks like @FIFAcom is on Twitter just in time to spread big news about the World Cup directly from South Africa!   

      country         date_time            id language  latitude  longitude  \
28263     NaN  26/09/2015 21:18  6.478830e+17      und       NaN        NaN   
28996     NaN  18/11/2016 08:34  7.995310e+17       en       NaN        NaN   
32910     NaN  23/10/2013 21:55  3.931340e+17       en       NaN        NaN   
3834      NaN  07/11/2015 02:41  6.628220e+17       en       NaN        NaN   
23410     NaN  09/06/2010 20:49  1.580143e+10       en       NaN        NaN   

       number_of_likes  number_of_shares        hashtags  \
28263             2891               700   [TheBFFIssue]   
28996             9121              3208  [SlumberParty]   
32910             8624             11800              []   
3834             44867             31307              []   
23410               56               297              []   

                            urls                                         temas  
28263   [http://t.co/VmVqbZ0OtZ]           [friendship, social media, fashion]  
28996                         []   [television, social media, event promotion]  
32910   [http://t.co/R2uUpf9o3x]               [football, sports, Real Madrid]  
3834   [https://t.co/nTj4mytOf9]  [social media, personal expression, brevity]  
23410                         []               [World Cup, FIFA, South Africa]

#### Exportacion del conjunto de datos con las variables expandidas

In [ ]:
df_completo.to_csv('/content/antropic2.csv', index=False)

In [ ]:
import pandas as pd
data=pd.read_csv('/content/antropic2.csv')

In [ ]:
data.head()

author  \
0  KimKardashian   
1  britneyspears   
2      Cristiano   
3   justinbieber   
4        Twitter   

                                                                                                                                        content  \
0                                                                               #TheBFFIssue @sorbetmag @riccardotisci17 http://t.co/VmVqbZ0OtZ   
1  Thank you to @mtv for premiering #SlumberParty on their official snapchat! I’ll be taking over their account tomorrow. Follow them!!! 👻: MTV   
2                                                                    What an exciting game at the Bernabéu! Hala Madrid! http://t.co/R2uUpf9o3x   
3                                                                                                           Mine too :) https://t.co/nTj4mytOf9   
4                             Looks like @FIFAcom is on Twitter just in time to spread big news about the World Cup directly from South Africa!   

  country         date_time            id language  latitude  longitude  \
0     NaN  26/09/2015 21:18  6.478830e+17      und       NaN        NaN   
1     NaN  18/11/2016 08:34  7.995310e+17       en       NaN        NaN   
2     NaN  23/10/2013 21:55  3.931340e+17       en       NaN        NaN   
3     NaN  07/11/2015 02:41  6.628220e+17       en       NaN        NaN   
4     NaN  09/06/2010 20:49  1.580143e+10       en       NaN        NaN   

   number_of_likes  number_of_shares          hashtags  \
0             2891               700   ['TheBFFIssue']   
1             9121              3208  ['SlumberParty']   
2             8624             11800                []   
3            44867             31307                []   
4               56               297                []   

                          urls  \
0   ['http://t.co/VmVqbZ0OtZ']   
1                           []   
2   ['http://t.co/R2uUpf9o3x']   
3  ['https://t.co/nTj4mytOf9']   
4                           []   

                                                temas  
0           ['friendship', 'social media', 'fashion']  
1   ['television', 'social media', 'event promotion']  
2               ['football', 'sports', 'Real Madrid']  
3  ['social media', 'personal expression', 'brevity']  
4               ['World Cup', 'FIFA', 'South Africa']

## **Analisis e Interpretacion de los resultados**